contracts/clients/test/OpertionResults.test.ts

```diff
@@ -0,0 +1,165 @@
+import { expect } from "chai";
+import { BigNumber, ContractReceipt, Event, utils } from "ethers";
+import * as TypeMoq from "typemoq";
+import { getOperationResults } from "../src";
+
+class MockResult extends Array<unknown> {
+  [key: string]: unknown;
+
+  constructor(obj: Record<string, unknown>) {
+    super();
+    for (const k in obj) {
+      this[k] = obj[k];
+    }
+  }
+}
+
+const getErrorResult = (actionIndex: number, message: string): string => {
+  const fullErrorMessage = `${actionIndex} - ${message}`;
+  const encodedErrorMessage = utils.defaultAbiCoder.encode(
+    ["string"],
+    [fullErrorMessage],
+  );
+  // Add empty methodId (4 bytes, 8 chars), remove leading 0
+  return `0x${"0".repeat(8)}${encodedErrorMessage.substring(2)}`;
+};
+
+describe("OperationResults", () => {
+  describe("getOperationResults", () => {
+    it("fails if no events are in transaction", () => {
+      const hash = "0xabc123";
+      const txnReceiptMock = TypeMoq.Mock.ofType<ContractReceipt>();
+      txnReceiptMock.setup((r) => r.transactionHash).returns(() => hash);
+      txnReceiptMock.setup((r) => r.events).returns(() => undefined);
+
+      expect(() => getOperationResults(txnReceiptMock.object)).to.throw(
+        `no events found in transaction ${hash}`,
+      );
```

Comment:
I'm concerned that `TypeMoq` is unnecessarily complex for our use case.

Why not just write:

```ts
const txnReceiptMoq = {
  transactionHash: hash,
  events: undefined,
} as ContractReceipt;
```

Reply:
I was trying to avoid `as` casts when possible as the underlying object technically fails proper typing. `typemoq` does add a bit of complexity but it will fully stub out the object, and could be useful in the future for testing more complex OO cases.

I will try without `typemoq`.

Reply:
> but it will fully stub out the object

That doesn't seem possible... on this line:
```ts
TypeMoq.Mock.ofType<ContractReceipt>()
```
`ContractReceipt` only exists in the type space, so the actual object produced here must not have any information about `ContractReceipt`.

Reply:
The stub defaults to `undefined` return values. There also is a behavior where you can make it throw if you don't define something. But I think at this point is is overkill so I just removed.